### PR TITLE
Refactor inc spec resolution to pass array properties

### LINKED ISSUE #320: Overridden dataset dtype and shape are not passed from parent to child
If a schema has two groups, e.g., `PatchClampSeries` and `CurrentClampSeries` defined as such:
```yaml
groups:
- neurodata_type_def: PatchClampSeries
  neurodata_type_inc: TimeSeries
  doc: ...
  datasets:
  - name: data
    dtype: numeric
    dims:
    - num_times
    shape:
    - null
    doc: ...
- neurodata_type_def: CurrentClampSeries
  neurodata_type_inc: PatchClampSeries
  doc: ...
  datasets:
  - name: data
    doc: Recorded voltage.
```
where the `'data'` dataset in `CurrentClampSeries` overrides the `'data'` dataset in the parent `PatchClampSeries`, then the `dtype`, `dims`, and `shape` keys of the parent `PatchClampSeries` `'data'` dataset are not inherited by the child `CurrentClampSeries` `'data'` dataset.

Is this intended behavior? @oruebel @ajtritt 

If so, the NWB schema needs to be amended in a few places.

### LINKED ISSUE #1311: [Bug]: Cannot build data (dataset) with dtype datetime
### What happened?

See https://github.com/nehatk17/ndx-multisubjects/issues/4

### Steps to Reproduce

```python
Create datasetspec that extends VectorData with dtype: datetime
```

### Traceback

```pytb
../hdmf/src/hdmf/utils.py:578: in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/backends/hdf5/h5tools.py:353: in write
    super().write(**kwargs)
../hdmf/src/hdmf/utils.py:578: in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/backends/io.py:98: in write
    f_builder = self.__manager.build(container, source=self.__source, root=True)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/utils.py:578: in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/build/manager.py:173: in build
    result = self.__type_map.build(container, self, source=source, spec_ext=spec_ext, export=export)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/utils.py:578: in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/build/manager.py:789: in build
    builder = obj_mapper.build(container, manager, builder=builder, source=source, spec_ext=spec_ext, export=export)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/utils.py:578: in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/build/objectmapper.py:765: in build
    self.__add_groups(builder, self.__spec.groups, container, manager, source, export)
../hdmf/src/hdmf/build/objectmapper.py:1164: in __add_groups
    self.__add_groups(sub_builder, spec.groups, container, build_manager, source, export)
../hdmf/src/hdmf/build/objectmapper.py:1177: in __add_groups
    self.__add_containers(builder, spec, attr_value, build_manager, source, container, export)
../hdmf/src/hdmf/build/objectmapper.py:1197: in __add_containers
    new_builder = build_manager.build(value, source=source, spec_ext=spec, export=export)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/utils.py:578: in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/build/manager.py:173: in build
    result = self.__type_map.build(container, self, source=source, spec_ext=spec_ext, export=export)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/utils.py:578: in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/build/manager.py:789: in build
    builder = obj_mapper.build(container, manager, builder=builder, source=source, spec_ext=spec_ext, export=export)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/utils.py:578: in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/build/objectmapper.py:764: in build
    self.__add_datasets(builder, self.__spec.datasets, container, manager, source, export)
../hdmf/src/hdmf/build/objectmapper.py:1147: in __add_datasets
    self.__add_containers(builder, spec, attr_value, build_manager, source, container, export)
../hdmf/src/hdmf/build/objectmapper.py:1197: in __add_containers
    new_builder = build_manager.build(value, source=source, spec_ext=spec, export=export)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/utils.py:578: in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/build/manager.py:173: in build
    result = self.__type_map.build(container, self, source=source, spec_ext=spec_ext, export=export)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/utils.py:578: in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/build/manager.py:789: in build
    builder = obj_mapper.build(container, manager, builder=builder, source=source, spec_ext=spec_ext, export=export)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../hdmf/src/hdmf/utils.py:578: in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
           ^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pynwb.io.core.VectorDataMap object at 0x10e8be850>
kwargs = {'builder': None, 'container': <hdmf.common.table.VectorData object at 0x10f88caa0>, 'export': False, 'manager': <hdmf.build.manager.BuildManager object at 0x10e8be350>, ...}
container = <hdmf.common.table.VectorData object at 0x10f88caa0>, manager = <hdmf.build.manager.BuildManager object at 0x10e8be350>, parent = None
source = '/Users/rly/Documents/NWB/ndx-multisubjects/test.nwb', builder = None
spec_ext = {'dtype': 'isodatetime', 'doc': 'Date of birth of subject. Can be supplied instead of age.', 'name': 'date_of_birth', 'quantity': '?', 'neurodata_type_inc': 'VectorData'}
export = False

    @docval({"name": "container", "type": AbstractContainer, "doc": "the container to convert to a Builder"},
            {"name": "manager", "type": BuildManager, "doc": "the BuildManager to use for managing this build"},
            {"name": "parent", "type": GroupBuilder, "doc": "the parent of the resulting Builder", 'default': None},
            {"name": "source", "type": str,
             "doc": "the source of container being built i.e. file path", 'default': None},
            {"name": "builder", "type": BaseBuilder, "doc": "the Builder to build on", 'default': None},
            {"name": "spec_ext", "type": BaseStorageSpec, "doc": "a spec extension", 'default': None},
            {"name": "export", "type": bool, "doc": "whether this build is for exporting",
             'default': False},
            returns="the Builder representing the given AbstractContainer", rtype=Builder)
    def build(self, **kwargs):
        '''Convert an AbstractContainer to a Builder representation.
    
        References are not added but are queued to be added in the BuildManager.
        '''
        container, manager, parent, source = getargs('container', 'manager', 'parent', 'source', kwargs)
        builder, spec_ext, export = getargs('builder', 'spec_ext', 'export', kwargs)
        name = manager.get_builder_name(container)
        if isinstance(self.__spec, GroupSpec):
            self.logger.debug("Building %s '%s' as a group (source: %s)"
                              % (container.__class__.__name__, container.name, repr(source)))
            if builder is None:
                builder = GroupBuilder(name, parent=parent, source=source)
            self.__add_datasets(builder, self.__spec.datasets, container, manager, source, export)
            self.__add_groups(builder, self.__spec.groups, container, manager, source, export)
            self.__add_links(builder, self.__spec.links, container, manager, source, export)
        else:
            if builder is None:
                if not isinstance(container, Data):
                    msg = "'container' must be of type Data with DatasetSpec"
                    raise ValueError(msg)
                spec_dtype, spec_shape, spec_dims, spec = self.__check_dset_spec(self.spec, spec_ext)
                dimension_labels = self.__get_dimension_labels_from_spec(container.data, spec_shape, spec_dims)
                if isinstance(spec_dtype, RefSpec):
                    self.logger.debug("Building %s '%s' as a dataset of references (source: %s)"
                                      % (container.__class__.__name__, container.name, repr(source)))
                    # create dataset builder with data=None as a placeholder. fill in with refs later
                    builder = DatasetBuilder(
                        name,
                        data=None,
                        parent=parent,
                        source=source,
                        dtype=spec_dtype.reftype,
                        dimension_labels=dimension_labels,
                    )
                    manager.queue_ref(self.__set_dataset_to_refs(builder, spec_dtype, spec_shape, container, manager))
                elif isinstance(spec_dtype, list):
                    # a compound dataset
                    self.logger.debug("Building %s '%s' as a dataset of compound dtypes (source: %s)"
                                      % (container.__class__.__name__, container.name, repr(source)))
                    # create dataset builder with data=None, dtype=None as a placeholder. fill in with refs later
                    builder = DatasetBuilder(
                        name,
                        data=None,
                        parent=parent,
                        source=source,
                        dtype=spec_dtype,
                        dimension_labels=dimension_labels,
                    )
                    manager.queue_ref(self.__set_compound_dataset_to_refs(builder, spec, spec_dtype, container,
                                                                          manager))
                else:
                    # a regular dtype
                    if spec_dtype is None and self.__is_reftype(container.data):
                        self.logger.debug("Building %s '%s' containing references as a dataset of unspecified dtype "
                                          "(source: %s)"
                                          % (container.__class__.__name__, container.name, repr(source)))
                        # an unspecified dtype and we were given references
                        # create dataset builder with data=None as a placeholder. fill in with refs later
                        builder = DatasetBuilder(
                            name,
                            data=None,
                            parent=parent,
                            source=source,
                            dtype="object",
                            dimension_labels=dimension_labels,
                        )
                        manager.queue_ref(self.__set_untyped_dataset_to_refs(builder, container, manager))
                    else:
                        # a dataset that has no references, pass the conversion off to the convert_dtype method
                        self.logger.debug("Building %s '%s' as a dataset (source: %s)"
                                          % (container.__class__.__name__, container.name, repr(source)))
                        try:
                            # use spec_dtype from self.spec when spec_ext does not specify dtype
                            if isinstance(container.data, TermSetWrapper):
                                data = container.data.value
                            else:
                                data = container.data
                            bldr_data, dtype = self.convert_dtype(spec, data, spec_dtype=spec_dtype)
                        except Exception as ex:
                            msg = f"could not resolve dtype for {type(container).__name__} '{container.name}'"
                            full_msg = f"{msg}: {str(ex)}"
>                           raise Exception(full_msg) from ex
E                           Exception: could not resolve dtype for VectorData 'date_of_birth': Expected unicode or ascii string, got <class 'datetime.date'>

../hdmf/src/hdmf/build/objectmapper.py:833: Exception
```

### Operating System

macOS

### Python Version

3.13

### Package Versions

latest dev version of hdmf

### PR DESCRIPTION
**Pull request by rly**
## Motivation
Fix #320

The key changes are:

1. __New resolution system__: Instead of resolving includes during spec loading, resolution now happens after all specs are loaded via `NamespaceCatalog.resolve_all_specs()`
2. __New methods__: `BaseStorageSpec.resolve_inc_spec()` replaces the old `BaseStorageSpec.resolve_spec()` method
3. __Resolution tracking__: New properties `BaseStorageSpec.resolved` and `BaseStorageSpec.inc_spec_resolved` track resolution state
4. __Cross-namespace resolution__: The system can now resolve specs that include types from different namespaces
5. `dtype`, `shape`, `dims`, `value`, and `default_value` in `DatasetSpec` and `AttributeSpec` are now inherited and validated from the parent data type spec
6. If `dims` are not provided in a `DatasetSpec` or `AttributeSpec`, but `shape` is provided, `dims` will be set to a list of dummy dimension names, e.g., "dim_0", "dim_1", etc.

This PR also adds a few TODOs for extra checks for cases that should not be allowed but are currently allowed (even before this PR). These will be handled in future PRs from #1316.

Supersedes / closes #321.

## How to test the behavior?
```
pytest
```

## Checklist

- [x] Did you update `CHANGELOG.md` with your changes?
- [x] Does the PR clearly describe the problem and the solution?
- [x] Have you reviewed our [Contributing Guide](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst)?
- [x] Does the PR use "Fix #XXX" notation to tell GitHub to close the relevant issue numbered XXX when the PR is merged?

### CODE REVIEW

**Comment by rly** [comment]
Complex spec resolution case:

- DynamicTable spec contains a VectorData spec for dataset named "col" with any shape, any dtype, no attributes
- ChildTable extends DynamicTable
- ChildTable specifies that the dataset named "col" should have 1D or 2D shape, dtype numeric, and some extra attribute attr1
- GrandchildTable extends ChildTable
- GrandchildTable specifies that the dataset named "col" should have 1D shape, dtype int, and some extra attribute attr2 (along with attr1)

Currently I think only the extra attributes are properly handled. See #320.

- ComplexGrandchildTable extends ChildTable
- ChildVectorData extends VectorData and says that the dataset should have attribute attr2
- ComplexGrandchildTable specifies that the dataset named "col" is actually a ChildVectorData type which says that the dataset should have 1D shape, datetime int, along with attr1 from ChildTable/col and attr2 from ChildVectorData

I don't think we correctly support specification or validation of this case.

**Comment by rly** [comment]
HDMF tests pass but I am running into an issue with NWB schema where NWBFile contains a LabMetaData, but LabMetaData is defined after NWBFile, so spec resolution fails. This did not happen before because spec resolution happened only for new type definitions. In this case, the group spec within LabMetaData
```
{'neurodata_type_inc': 'LabMetaData', 'doc': 'Place-holder than can be extended so that lab-specific meta-data can be placed in /general.', 'quantity': '*'}
```
does not have a neurodata_type_def. Previously the spec would not be resolved (so if this were a dataset, then any dtype/shape specifications in LabMetaData would not be inherited).

**Comment by rly** [comment]
Thoughts to self about extra fields: A TimeSeries spec adds groups, datasets, etc. to a Container spec. A TimeSeries object should also conform to the Container spec. This means validation of an object against the Container spec allows for extra fields. This doesn't really come up because I think right now, validation happens only against the most specific data type (only TimeSeries, not also Container) with its fully resolved spec (all inherited specs from parent types have been copied/refined into the most specific type spec). (Alternatively, Container and most other types could be updated to allow any number of arbitrary groups, datasets, links, and attributes, but this changes everything and is not intuitive.)

See also discussion about whether fields can be added to groups/dsets with only `data_type_inc`: https://github.com/hdmf-dev/hdmf-schema-language/issues/13

**Comment by oruebel** [comment]
> This means validation of an object against the Container spec allows for extra fields

Agreed. This is consistent with standard inheritance in that `Container` expects all the fields that it defines to be present but it doesn't care about fields that a child type might have added.

**Comment by rly** [comment]
Related: 
- #1086
- #1213

**Comment by rly** [comment]
In https://github.com/hdmf-dev/hdmf/issues/1086 we discussed that resolving the order of type definitions for resolving included/linked specs for all cases would be quite complex, so we should try to cover the most common use cases (notably the ones in the core NWB schema) and recommend a best practice for defining types before using them.

**Comment by rly** [comment]
Note: this warning now shows up:
```
tests/unit/test_io_hdf5_h5tools.py::TestCpdDatasetRefs::test_roundtrip
  /Users/rly/Documents/NWB/hdmf/src/hdmf/build/objectmapper.py:911: IncorrectDatasetShapeBuildWarning: Shape of data does not match shape in spec 'baz_cpd_data'
    warnings.warn(msg, IncorrectDatasetShapeBuildWarning)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
```

type: BazCpdData which extends Data
data = a list of 10 (integer, object) pairs 
spec = compound dtype (integer, object reference), with 1D shape

Same problem as https://github.com/hdmf-dev/hdmf/issues/1191 but now surfaced (as it should while there is not yet a fix).

It's not clear to me why it shows up now. Will have to investigate further.

**Comment by rly** [comment]
> It's not clear to me why it shows up now. Will have to investigate further.

This is because I made a change to `DatasetSpec.__init__` to create dummy values (e.g., `dim_0`, `dim_1`, etc.) for `dims` if `shape` is provided and `dims` is not (`xarray` also does this). This change was necessary to cover the case where a new data type `MyColumn` extends `VectorData` (which defines both `shape` and `dims`) but refines only `shape`. As a result, `shape` was not consistent with the inherited `dims`. With this change, `MyColumn.dims` is set to dummy values that match `MyColumn.shape` unless explicitly specified. For simplicity, I think HDMF should create these dummy values for `dims` if `shape` is provided and `dims` is not, to keep them consistent and remove this edge case. 

TODO:
- I don't think the complex spec resolution case 1 above is fixed yet. Write a test.
- Add more tests for all the changes

This PR contains breaking changes that will be summarized in the changelog.
- GroupSpec.data_type_inc and DatasetSpec.data_type_inc no longer accept a spec
- GroupSpec and DatasetSpec no longer automatically resolve their includes after construction. You must call `spec.resolve_inc_spec(inc_spec)` or `namespace_catalog.resolve_all_specs()` if they are in a namespace catalog
- dtype/shape/dims compatibility is now checked for datasets and attributes against the included type
- validation of extension dtype/shape/dims should now happen

File: tests/unit/build_tests/test_classgenerator.py [hunk 1]
```diff
@@ -732,8 +753,10 @@ def _build_separate_namespaces(self):
             data_type_def='Baz',
             data_type_inc='Bar',
             groups=[
-                GroupSpec(data_type_inc='Qux', doc='a qux', quantity='?'),
-                GroupSpec(data_type_inc='Bar', doc='a bar', quantity='?')
+                GroupSpec(data_type_inc='Bar', doc='a bar', quantity='?'),
+            ],
+            datasets=[
+                DatasetSpec(data_type_inc='Qux', doc='a qux', quantity='?'),
```
[the remark this answers is not in the document]
Comment (rly) [reply tests/unit/build_tests/test_classgenerator.py:756]: This is intentional. Qux is supposed to be a dataset

File: src/hdmf/spec/spec.py [hunk 2]
```diff
@@ -77,6 +78,136 @@ def check_dtype(dtype):
...
+def _get_dtype_name_prec_level(dtype: str):
+    # TODO: this returns just the first letter of the dtype name, which is not ideal
+    # TODO: this does not handle the "numeric" dtype case. use hdmf.validate.validator.__allowable
+    # TODO: not all dtype names have numbers in them that indicate precision level
+    m = re.search('[0-9]+', dtype)
+    if m is not None:
+        prec = int(m.group())
+    else:
+        prec = 32
+    return dtype[0], prec
```
[the remark this answers is not in the document]
Comment (rly) [reply src/hdmf/spec/spec.py]: This will be handled in a separate PR.

File: src/hdmf/spec/spec.py [hunk 3]
```diff
@@ -77,6 +78,136 @@ def check_dtype(dtype):
...
+
+
+def _is_sub_dtype(new: Union[str, "RefSpec"], orig: Union[str, "RefSpec"]):
+    if isinstance(orig, RefSpec):
+        if not isinstance(new, RefSpec):
+            return False
+        # check ref target is a subtype of the original ref target
+        # TODO: implement subtype check for RefSpec. might need to resolve RefSpec target type to a spec first
+        # return orig == new
+        return True
```
[the remark this answers is not in the document]
Comment (rly) [reply src/hdmf/spec/spec.py:122]: This is incorrect and will be handled in a separate PR.

**Comment by rly** [comment]
Hi @stephprince , this is a big one to review. Sorry there are so many changes. Let me know if you would like to meet to discuss.

File: src/hdmf/spec/namespace.py [hunk 4]
```diff
@@ -426,25 +437,71 @@ def __convert_spec_cls_keys(self, parent_cls, spec_cls, spec_dict):
...
+        # iteratively resolve specs until all are resolved or we reach max passes
+        num_passes = 0
+        max_passes = 10
+        while unresolved_data_type_specs and num_passes < max_passes:
+            num_passes += 1
+            specs_to_process = copy(unresolved_data_type_specs)
+            for spec in specs_to_process:
+                self.__resolve_spec(spec)
+                if spec.resolved:
+                    unresolved_data_type_specs.remove(spec)
```
Comment (rly) [inline src/hdmf/spec/namespace.py]: Comments from discussion with @stephprince. The graph might be a little clearer to understand and more efficient. This involves more code though.

**Comment by rly** [comment]
One extension test fails with this branch: ndx-sound. I raised https://github.com/catalystneuro/ndx-sound/issues/14

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/ndx_sound/__init__.py", line 24, in <module>
    load_namespaces(ndx_sound_specpath)
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/hdmf/utils.py", line 582, in func_call
    return func(**pargs)
           ^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/pynwb/__init__.py", line 165, in load_namespaces
    return __TYPE_MAP.load_namespaces(namespace_path)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/hdmf/utils.py", line 578, in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/hdmf/build/manager.py", line 483, in load_namespaces
    deps = self.__ns_catalog.load_namespaces(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/hdmf/utils.py", line 578, in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/hdmf/spec/namespace.py", line 664, in load_namespaces
    self.resolve_all_specs()
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/hdmf/spec/namespace.py", line 456, in resolve_all_specs
    self.__resolve_namespace_specs(namespace)
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/hdmf/spec/namespace.py", line 516, in __resolve_namespace_specs
    __resolve_local(spec)
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/hdmf/spec/namespace.py", line 500, in __resolve_local
    spec.resolve_inc_spec(included_spec)
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/hdmf/utils.py", line 578, in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/hdmf/spec/spec.py", line 1044, in resolve_inc_spec
    self.__datasets[dataset.name].resolve_inc_spec(dataset)
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/hdmf/utils.py", line 578, in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/hdmf/spec/spec.py", line 868, in resolve_inc_spec
    _resolve_inc_spec_dtype(self, inc_spec)
  File "/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/hdmf/spec/spec.py", line 145, in _resolve_inc_spec_dtype
    raise ValueError(msg)
ValueError: Cannot extend float64 to float
```

File: src/hdmf/spec/namespace.py [hunk 5]
```diff
@@ -248,8 +249,6 @@ def __init__(self, **kwargs):
         self.__included_specs = dict()
         self.__included_sources = dict()
 
-        self._loaded_specs = self.__loaded_specs
```
Comment (rly) [inline src/hdmf/spec/namespace.py:251]: `self._loaded_specs` is unused so I removed it.

File: tests/unit/spec_tests/test_dataset_spec.py [hunk 6]
```diff
@@ -73,40 +94,6 @@ def test_constructor_ref_spec(self):
                            data_type_def='EphysData')
         self.assertDictEqual(spec['dtype'], dtype)
 
-    def test_datatype_extension(self):
```
Comment (rly) [inline tests/unit/spec_tests/test_dataset_spec.py:76]: Spec resolution tests were moved to test_spec_resolution.py and expanded upon

**Comment by rly** [comment]
@stephprince This is now ready for review. I refactored `resolve_inc_spec` to take a `SpecNamespace` so that we can check whether a `data_type_inc` in a subspec is compatible with the included (inherited) `data_type_inc`.

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,20 @@
...
+  - New methods: `BaseStorageSpec.resolve_inc_spec()` replaces the old `BaseStorageSpec.resolve_spec()` method
+  - Resolution tracking: New properties `BaseStorageSpec.resolved` and `BaseStorageSpec.inc_spec_resolved` track resolution state
+  - Cross-namespace resolution: The system can now resolve specs that include types from different namespaces
+  - `dtype`, `shape`, `dims`, `value`, and `default_value` in `DatasetSpec` and `AttributeSpec` are now inherited and validated from the parent data type spec
+- If `dims` are not provided in a `DatasetSpec` or `AttributeSpec`, but `shape` is provided, `dims` will be set to a list of dummy dimension names, e.g., "dim_0", "dim_1", etc. @rly [#1312](https://github.com/hdmf-dev/hdmf/pull/1312)
+
+### Added
+- Warning when `data_type_def` and `data_type_inc` are the same in a spec. @rly [#1312](https://github.com/hdmf-dev/hdmf/pull/1312)
+- New methods `NamespaceCatalog.get_namespace_for_type()` and `NamespaceCatalog.get_spec_for_type()` to retrieve the namespace or spec for a given data type. @rly [#1312](https://github.com/hdmf-dev/hdmf/pull/1312)
```
Comment (stephprince) [inline CHANGELOG.md]:
```suggestion
```

These ended up being removed while you were iterating on this PR yes? I don't see them anymore

**Review by stephprince — APPROVED** [review]
Looks good to me @rly! Sorry for the delay.

File: CHANGELOG.md (hunk 7 above)
Comment (rly) [reply CHANGELOG.md]: You're right! Thank you
